### PR TITLE
Indexes::search() improvements

### PR DESCRIPTION
**Pull request by Guikingone**
Hi everyone 👋 

As mentioned here: https://github.com/meilisearch/meilisearch-php/issues/119, here's the `SearchResult` class and the improvements that comes with it. 

There's just a small "DX" way that I'm not sure about: 

- When using `SearchResult::filter()` should the `hits` array be reset to 0 or stay as it? 

IMO, the array could stays as it as we return a new `SearchResult` object but sometimes, it could be good idea to reset the keys from 0 (loops, etc) 🤔 

Thanks for the feedback and have a great day 🙂

### CODE REVIEW

**Comment by curquiza** [comment]
Hello @Guikingone! So sorry for the delay!!!
The integration team is mostly ok with this change! 😄
I just need a final discussion with the developer relation manager before reviewing and merging this PR: indeed, we would apply this change in other SDKs (ruby, python...), that's why I want to be sure before taking this kind of decision that differs from Algolia!
Thanks again for your patience and your suggestion -> that's exactly the kind of solution I was looking for to help users managing their search results!

**Comment by Guikingone** [comment]
Hello @curquiza 👋 

The checks are green after a small refactoring, feel free to review it when you have time, I've introduced the `json()` method as discussed 🙂

**Comment by curquiza** [comment]
Thank you so much @Guikingone!
I review it asap! 👍

**Review by ppshobi — APPROVED** [review]
Seems good to me. Real nice touch with the tests

**Review by curquiza — CHANGES_REQUESTED** [review]
Thanks for this huge job, I just ask for small modifications!

About your question, yes let it as it is for the moment, it seems ok for me 🙂

Again, thank you so much for your work 🙏 Be sure we are going to extend this way of handling search result to our other SDKs.

FYI: I'm not going to release right after merging this PR. I indeed need to discuss naming with my team, and most of all, another breaking PR is planned so I will do the push to Packagist after all the breaking changes are done 😇

File: src/Search/SearchResult.php [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+            'nbHits' => \count($this->hits),
+            'exhaustiveNbHits' => $this->exhaustiveNbHits,
+            'processingTimeMs' => $this->processingTimeMs,
+            'query' => $this->query,
+            'exhaustiveFacetsCount' => $this->exhaustiveFacetsCount,
+            'facetsDistribution' => $this->facetsDistribution,
+        ];
+    }
+
+    public function json(): string
```
[the remark this answers is not in the document]
Comment (Guikingone) [reply src/Search/SearchResult.php:184]: Fixed

File: src/Search/SearchResult.php [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    }
+
+    public function toArray(): array
+    {
+        return [
+            'hits' => $this->hits,
+            'offset' => $this->offset,
+            'limit' => $this->limit,
+            'matches' => $this->nbHits,
+            'nbHits' => \count($this->hits),
```
[the remark this answers is not in the document]
Comment (Guikingone) [reply src/Search/SearchResult.php:175]: Fixed

File: src/Endpoints/Indexes.php [hunk 3]
```diff
@@ -117,14 +118,31 @@ public function waitForPendingUpdate($updateId, $timeoutInMs = 5000, $intervalIn
...
+        $result = $this->http->post(self::PATH.'/'.$this->uid.'/search', $parameters);
+
+        if (\array_key_exists('raw', $options) && $options['raw']) {
+            return $result;
+        }
+
+        $searchResult = new SearchResult($result);
+
+        if (\array_key_exists('filteringHits', $options) && \is_callable($options['filteringHits'])) {
+            $searchResult = $searchResult->filter($options['filteringHits']);
```
[the remark this answers is not in the document]
Comment (Guikingone) [reply src/Endpoints/Indexes.php:142]: Improved 🙂

**Review by curquiza — APPROVED** [review]
Finally merging! Thanks a lot @Guikingone and also @ppshobi! 🎉 

bors merge

**Pull request merged**